src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs

```diff
@@ -436,5 +438,123 @@ public DummyTypeInfo GetTypeFromSpecification(MetadataReader reader, ModuleToken
+            knownTypes.Add(valueTask);
+            knownMethods.Add(valueTask.GetKnownMethod("FromResult"u8, null));
+            knownMethods.Add(valueTask.GetKnownMethod("get_CompletedTask"u8, null));
+            knownMethods.Add(valueTask.GetKnownMethod("get_IsCompleted"u8, null));
+            knownMethods.Add(valueTask.GetKnownMethod("ThrowIfCompletedUnsuccessfully"u8, null));
+            knownMethods.Add(valueTask.GetKnownMethod("AsTaskOrNotifier"u8, null));
+
+            var valueTaskGeneric = CompilerContext.SystemModule.GetKnownType("System.Threading.Tasks"u8, "ValueTask`1"u8);
+            knownTypes.Add(valueTaskGeneric);
+            knownMethods.Add(valueTaskGeneric.GetKnownMethod("get_Result"u8, null));
+            knownMethods.Add(valueTaskGeneric.GetKnownMethod("AsTaskOrNotifier"u8, null));
+
+            KnownTypes = knownTypes.ToImmutable();
+            KnownMethods = knownMethods.ToImmutable();
+
+            try
+            {
+                Debug.Assert(_manifestMutableModule.ModuleThatIsCurrentlyTheSourceOfNewReferences == null);
+                _manifestMutableModule.ModuleThatIsCurrentlyTheSourceOfNewReferences = _manifestMutableModule.Context.SystemModule;
+                _manifestMutableModule.AddingReferencesToR2RKnownTypesAndMethods = true;
+                foreach (var type in KnownTypes)
+                {
+                    // We can skip types that are already in the version bubble
+                    if (_compilationModuleGroup.VersionsWithType(type))
+                        continue;
+
+                    EntityHandle? handle = _manifestMutableModule.TryGetEntityHandle(type);
+                    if (!handle.HasValue)
+                        throw new NotImplementedException($"Entity handle for known type ({type}) not found in manifest module");
+
+                    var token = new ModuleToken(_manifestMutableModule, handle.Value);
+                    this.AddModuleTokenForType(type, token);
+                    Debug.Assert(!default(ModuleToken).Equals(this.GetModuleTokenForType((EcmaType)type, allowDynamicallyCreatedReference: true, throwIfNotFound: true)));
+                }
+
+                foreach (var method in KnownMethods)
+                {
+                    // We can skip methods that are already in the version bubble
+                    if (_compilationModuleGroup.VersionsWithMethodBody((EcmaMethod)method))
+                        continue;
+
+                    EntityHandle ? handle = _manifestMutableModule.TryGetEntityHandle(method);
+                    if (!handle.HasValue)
+                        throw new NotImplementedException($"Entity handle for known method ({method}) not found in manifest module");
+
+                    var token = new ModuleToken(_manifestMutableModule, handle.Value);
+                    this.AddModuleTokenForMethod(method, token);
```

Comment:
Does this mean that every R2R image is going to have metadata for references to all async helpers? 

We should be creating these metadata tokens lazily, only when the async helper is actually used.